sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/coders/AwsCoders.java

```diff
@@ -0,0 +1,138 @@
+    return ResponseMetadataCoder.of();
+  }
+
+  /**
+   * Returns a new coder for SdkHttpMetadata.
+   *
+   * @return the SdkHttpMetadata coder
+   */
+  public static Coder<SdkHttpMetadata> sdkHttpMetadata() {
+    return new SdkHttpMetadataCoder(true);
+  }
+
+  /**
+   * Returns a new coder for SdkHttpMetadata that does not serialize the response headers.
+   *
+   * @return the SdkHttpMetadata coder
+   */
+  public static Coder<SdkHttpMetadata> sdkHttpMetadataWithoutHeaders() {
+    return new SdkHttpMetadataCoder(false);
+  }
+
+  private static class ResponseMetadataCoder extends AtomicCoder<ResponseMetadata> {
+
+    private static final Coder<Map<String, String>> METADATA_ENCODER =
+        NullableCoder.of(MapCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()));
+    private static final ResponseMetadataCoder INSTANCE = new ResponseMetadataCoder();
+
+    private ResponseMetadataCoder() {}
+
+    public static ResponseMetadataCoder of() {
+      return INSTANCE;
+    }
+
+    @Override
+    public void encode(ResponseMetadata value, OutputStream outStream)
+        throws CoderException, IOException {
+      METADATA_ENCODER.encode(
+          ImmutableMap.of(ResponseMetadata.AWS_REQUEST_ID, value.getRequestId()), outStream);
+    }
+
+    @Override
+    public ResponseMetadata decode(InputStream inStream) throws CoderException, IOException {
+      return new ResponseMetadata(METADATA_ENCODER.decode(inStream));
+    }
+  }
+
+  private static class SdkHttpMetadataCoder extends CustomCoder<SdkHttpMetadata> {
+
+    private static final Coder<Integer> STATUS_CODE_CODER = VarIntCoder.of();
+    private static final Coder<Map<String, String>> HEADERS_ENCODER =
+        NullableCoder.of(MapCoder.of(StringUtf8Coder.of(), StringUtf8Coder.of()));
+
+    private final boolean includeHeaders;
+
+    protected SdkHttpMetadataCoder(boolean includeHeaders) {
+      this.includeHeaders = includeHeaders;
+    }
+
+    @Override
+    public void encode(SdkHttpMetadata value, OutputStream outStream)
+        throws CoderException, IOException {
+      STATUS_CODE_CODER.encode(value.getHttpStatusCode(), outStream);
+      if (includeHeaders) {
+        HEADERS_ENCODER.encode(value.getHttpHeaders(), outStream);
+      }
+    }
+
+    @Override
+    public SdkHttpMetadata decode(InputStream inStream) throws CoderException, IOException {
+      final int httpStatusCode = STATUS_CODE_CODER.decode(inStream);
+      HttpResponse httpResponse = new HttpResponse(null, null);
+      httpResponse.setStatusCode(httpStatusCode);
+      if (includeHeaders) {
+        Optional.ofNullable(HEADERS_ENCODER.decode(inStream))
+            .ifPresent(
+                headers ->
+                    headers.keySet().forEach(k -> httpResponse.addHeader(k, headers.get(k))));
+      }
+      return SdkHttpMetadata.from(httpResponse);
+    }
+
+    @Override
+    public void verifyDeterministic() throws NonDeterministicException {
```

Comment:
And one more curiosity, why we have to implement `verifyDeterministic` for this coder?

Reply:
By default coders are considered non-deterministic which prevent them from being used as the key coder for many things (GBK, map/multimap side input keys, map state keys, ...).

Implementing it allows us to be deterministic if the coders we use are deterministic. So if the internally used coders become deterministic then we get that for free.

Reply:
`CustomCoder`'s implementation always throws NonDeterministicException so this seemed more correct.
https://github.com/apache/beam/blob/33ec0bb052e6711efaa833d6c044680e4de30902/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CustomCoder.java#L47-L53